### DISCUSSION TITLE
[Tutorial] Way to get Radio Messages

### DISCUSSION THREAD

**Discussion by siddharth-tewari** [theOehrly/Fast-F1#84] [Show and tell]
If found this so creating a issue so everyone gets to know about it.

1. Go to https://livetiming.formula1.com/static/SessionInfo.json
2. Copy the Url for the latest session.
Example of URL - 2021/2021-10-24_United_States_Grand_Prix/2021-10-24_Race/
3. Add prefix "livetiming.formula1.com/static" to the above url.
Example - livetiming.formula1.com/static/2021/2021-10-24_United_States_Grand_Prix/2021-10-24_Race/
4. Add TeamRadio.jsonStream to the end of the url to get links.
Example - livetiming.formula1.com/static/2021/2021-10-24_United_States_Grand_Prix/2021-10-24_Race/TeamRadio.jsonStream
5. Open the file in text editor and paste the links as suffix.
Example - http://livetiming.formula1.com/static/2021/2021-04-18_Emilia_Romagna_Grand_Prix/2021-04-18_Race/TeamRadio/MAXVER01_33_20210418_162030.mp3
6. Enjoy listening to recorded radio messages

**Comment by siddharth-tewari** [comment]
@theOehrly Maybe you can add it to the package to output the links directly in a row.

**Comment by theOehrly** [comment]
I could add it in theory, yes. But there are way too many things currently that I consider to be more important than adding this feature. So it will likely not happen any time in the foreseeable future.

**Comment by siddharth-tewari** [comment]
Hello, made a program for the same. It returns the radio messages for the latest GP. 
https://github.com/siddharth-tewari/F1-TeamRadio

**Comment by siddharth-tewari** [comment]
@theOehrly Maybe you can check my code and add it using by using that file in fast f1

**Comment by hash55** [comment]
Just a quick question. Does this work to get RaceControl messages as well? Apart from the radio mesages

**Comment by siddharth-tewari** [comment]
@hash55 It only gives the radio message with which Constructor's want to and rest is for F1TV. You can add a function and copy the code of radio and replace "TeamRadio.json" with "TrackStatus.json" for SC/VSC/Flags etc. rest all is not available on the api according to me.

**Comment by hash55** [comment]
Thanks! However, the official F1App also gives race control messages in the same window as the team radios. (I'm talking about messages like "lap deleted" or "pit lane closed" etc) so I would say it must be somewhere in the API, but I'm not familiar to them. I think it should be interesting to  have access to it, but I don't know where to start searching it XD

**Comment by theOehrly** [comment]
> Thanks! However, the official F1App also gives race control messages in the same window as the team radios. (I'm talking about messages like "lap deleted" or "pit lane closed" etc) so I would say it must be somewhere in the API, but I'm not familiar to them. I think it should be interesting to have access to it, but I don't know where to start searching it XD

At the top of 'fastf1/api.py' there is a dictionary called `pages` which contains known api endpoints. Even non-implemented ones. 
'RaceControlMessages.json' seems like the correct extension for the api base path of each event (`Session.api_path`).
You'll need to download the data manually or write yourself some code to do it.

**Comment by hash55** [comment]
Thanks! Will have a look at it

**Comment by siddharth-tewari** [comment]
@theOehrly I also took it from there only. Thank You! for making this list

**Comment by siddharth-tewari** [comment]
@theOehrly @hash55 Hello, try this code.
```
import urllib.request as request
import json
import requests
from jsonstream import load
from io import StringIO
import logging

def radio():
    print("Team Radio")
    with request.urlopen('http://livetiming.formula1.com/static/SessionInfo.json') as response:
        if response.getcode() == 200:
                logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
                logging.info('Getting the URL')
                source = response.read()
                data = json.loads(source)
                location = "http://livetiming.formula1.com/static/" + data['Path'] + "TeamRadio.json"
                logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
                logging.info('URL Successfully Found')
                url = location
                print(url)
                r = requests.get(location)
                r.encoding = 'utf-8-sig'
                out = json.loads(r.text)
                print(json.dumps(out, indent = 4))
                output = json.dumps(out, indent = 4)
                
        else:
            print('Unknown Error Occured')


def racecontrol():
    print("Race Control")
    with request.urlopen('http://livetiming.formula1.com/static/SessionInfo.json') as response:
        if response.getcode() == 200:
                logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
                logging.info('Getting the URL')
                source = response.read()
                data = json.loads(source)
                location = "http://livetiming.formula1.com/static/" + data['Path'] + "RaceControlMessages.json"
                logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
                logging.info('URL Successfully Found')
                url = location
                print(url)
                r = requests.get(location)
                r.encoding = 'utf-8-sig'
                out = json.loads(r.text)
                print(json.dumps(out, indent = 4))
                output = json.dumps(out, indent = 4)
                
        else:
            print('Unknown Error Occured')


def list():
    print("Driver List")
    with request.urlopen('http://livetiming.formula1.com/static/SessionInfo.json') as response:
        if response.getcode() == 200:
                logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
                logging.info('Getting the URL')
                source = response.read()
                data = json.loads(source)
                location = "http://livetiming.formula1.com/static/" + data['Path'] + "DriverList.json"
                logging.basicConfig(format='%(asctime)s - %(message)s', level=logging.INFO)
                logging.info('URL Successfully Found')
                url = location
                print(url)
                r = requests.get(location)
                r.encoding = 'utf-8-sig'
                out = json.loads(r.text)
                print(json.dumps(out, indent = 4))
                output = json.dumps(out, indent = 4)
                
        else:
            print('Unknown Error Occured')




radio()
racecontrol()
list()
```

**Comment by hash55** [comment]
Thanks! Works perfect, just had to change `from jsonstream import load`  to `from json import load`. But thanks, that is what I wanted

**Comment by siddharth-tewari** [comment]
@hash55 Actually Jsonstream is a package, you can install it using ```pip install jsonstream```, then the code would work faster using that package.

**Comment by hash55** [comment]
Oh ok. Thanks again. Btw in the `pages` section of the documentation, do you know what `Heartbeat.json` does? Looks like it's empty

**Comment by siddharth-tewari** [comment]
@hash55 Maybe it helps in checking if data if live during transmission and matches the sequence of other packets of data uploaded.

**Comment by theOehrly** [comment]
@hash55 well, it's not empty, but it only contains timestamps in 15 second intervals. I'm not sure that they're really useful for anything here. But I guess it's useful during the live data stream.

**Comment by hash55** [comment]
Thanks! Might there be more pages than those that are in the `pages `section? Or do you think you have discovered all of them?

**Comment by theOehrly** [comment]
Of course, I could have missed something. Also, I haven't checked in quite some time. So maybe I'll do that again in the next few days.

**Comment by siddharth-tewari** [comment]
@theOehrly I have a question that how do you find the pages from the api?

**Comment by hash55** [comment]
@siddharth-tewari  that's a good question. Let me know if you discover something new

**Comment by theOehrly** [comment]
There are different ways (I guess) how you can check. I don't really want to detail this here. Nothing special to this api but just "the usual way" of how you'd approach something like this.

I found one more api endpoint: "SessionData.json". It's not very interesting, though. It's simply a combination of session status, track status and lap count. So no new data. Still, I'll add it to the list of pages.

**Comment by siddharth-tewari** [comment]
@theOehrly Ok, actually I know nothing about API’s so can you please some reference links?
